pkg/monitors/nagios/nagios.go

```diff
@@ -0,0 +1,253 @@
+	var killTimer *time.Timer
+
+	// Parse command string with args
+	splitCmd, err := shellquote.Split(command)
+	if err != nil || len(splitCmd) == 0 {
+		return nil, nil, fmt.Errorf("exec: unable to parse command, %s", err)
+	}
+	// Prepare command exec
+	cmd := exec.Command(splitCmd[0], splitCmd[1:]...)
+	cmd.Stdout = &cmdOut
+	cmd.Stderr = &cmdErr
+	// Start command
+	if err = cmd.Start(); err != nil {
+		return
+	}
+
+	// After user timeout, try to gracefully stop the process with sigterm
+	termTimer := time.AfterFunc(time.Duration(termTimeout)*time.Second, func() {
+		err = cmd.Process.Signal(syscall.SIGTERM)
+		if err != nil {
+			return
+		}
+		// Kill the process to respect the monitor intervalSeconds
+		killTimer = time.AfterFunc(time.Duration(killTimeout)*time.Second, func() {
+			err = cmd.Process.Signal(syscall.SIGTERM)
+			if err != nil {
+				return
+			}
+		})
+	})
+	err = cmd.Wait()
+	// Shutdown timers
+	if killTimer != nil {
+		killTimer.Stop()
+	}
+	// Returns timeout error if sigterm was sent
+	if !termTimer.Stop() {
+		err = fmt.Errorf("timeout reached, %s", err)
+		return
+	}
+	stdout = cmdOut.Bytes()
+	stderr = cmdErr.Bytes()
+	// Returns exec related error
+	if err != nil {
+		return
+	}
+	// Returns success
+	return stdout, stderr, err
+}
+
+func getExitCode(err error, stdout []byte) (int, error) {
+	// See https://nagios-plugins.org/doc/guidelines.html#AEN78
+	// Some scripts could not respect the nagios convention and
+	// returns a code == 0 even if it show an error message.
+	// To mitigate this risk we override an anormal state when
+	// information allow to think the exit code is not relevant
+	if err == nil {
+		// We only need the first 3 bytes to determine the status
+		// from the string output (over the exit code)
+		status := strings.ToLower(string(stdout[:4]))
+		switch status {
+		case "crit":
+			return 2, nil
+		case "warn":
+			return 1, nil
+		case "unkn":
+			return unknown, nil
+		}
+		return 0, nil
+	}
+	// If there is an error but we cannot get the exit code so
+	// we cannot really consider it is a "critical" or a "warning"
+	// and we will use the "unknown" state
+	ee, ok := err.(*exec.ExitError)
+	if !ok {
+		// If the error is not an ExitError so the command could have been
+		// killed because of timeout
+		if strings.HasPrefix(err.Error(), "timeout") {
+			return unknown, err
+		}
+		return unknown, fmt.Errorf("command ended unexpectedly %s", err)
+	}
+
+	ws, ok := ee.Sys().(syscall.WaitStatus)
+	if !ok {
+		return 3, fmt.Errorf("cannot get exit code")
+	}
+
+	// An error with the exit code so we can be sure the command has been
+	// executed to its end even if the exitcode still can be an error in
+	// the script (not the nagios state). In any case we return the exit
+	// code as state and the user could find more information in the event
+	return ws.ExitStatus(), nil
+}
+
+func makeProperties(state int, err error, stdout []byte, stderr []byte) (map[string]interface{}, map[string]interface{}) {
```

Comment:
I would just make this function return the map of all of the properties and then have another function that filters out what you are omitting from `diffProperties`.  That will make it a bit clearer to read this in the main monitor method that calls this.

Reply:
you are right